File: java/src/main/java/com/azure/data/cosmos/serialization/hybridrow/schemas/ScopePropertyType.java
// Copyright (c) Microsoft Corporation. All rights reserved.
// Licensed under the MIT License.

package com.azure.data.cosmos.serialization.hybridrow.schemas;

public abstract class ScopePropertyType extends PropertyType {

    private boolean immutable;

    /**
     * True if the property's child elements cannot be mutated in place.
     * Immutable properties can still be replaced in their entirety.
     */
    public final boolean immutable() {
        return this.immutable;
    }

    public final void immutable(boolean value) {
        this.immutable = value;
    }
}

File: src/Serialization/HybridRow/Schemas/ScopePropertyType.cs
// ------------------------------------------------------------
//  Copyright (c) Microsoft Corporation.  All rights reserved.
// ------------------------------------------------------------

namespace Microsoft.Azure.Cosmos.Serialization.HybridRow.Schemas
{
    using Newtonsoft.Json;

    public abstract class ScopePropertyType : PropertyType
    {
        protected ScopePropertyType(TypeKind type) : base(type)
        {
        }

        /// <summary>True if the property's child elements cannot be mutated in place.</summary>
        /// <remarks>Immutable properties can still be replaced in their entirety.</remarks>
        [JsonProperty(PropertyName = "immutable")]
        [JsonConverter(typeof(StrictBooleanConverter))]
        public bool Immutable { get; set; }
    }
}
